compiler/src/dotty/tools/dotc/transform/SyntheticMembers.scala

```diff
@@ -194,6 +196,33 @@ class SyntheticMembers(thisPhase: DenotTransformer) {
       Match(index, (cases :+ generateIOBECase(index)).toList)
     }
 
+    /** The class
+     *
+     *  ```
+     *  case class C(x: T, y: T)
+     *  ```
+     *
+     *  gets the `productElement` method:
+     *
+     *  ```
+     *  def productElement(index: Int): Any = index match {
+     *    case 0 => this.x
+     *    case 1 => this.y
+     *    case _ => throw new IndexOutOfBoundsException(index.toString)
+     *  }
+     *  ```
+     */
+    def productElementBodyForScala2Compat(arity: Int, index: Tree)(using Context): Tree = {
+      val paramNames = ctx.owner.owner.info.decls.toList.filter(_.flags.is(ParamAccessor)).map(_.name)
+      // case N => this.${paramNames(N)}
+      val cases = paramNames.zip(0.until(arity)).map { (name, i) =>
+        val sel = This(clazz).select(name, _.info.isParameterless)
```

Comment:
Why not directly select the symbol found above instead?

Reply:
I followed the previous implementation too closely. Now I am selecting the symbols directly. I also used `zipWithIndex` instead of the zip with the range.